### ISSUE TITLE
Preserve oldest snapshot when forget policies not satisfied

### ISSUE THREAD

**Issue by kapitainsky** [rustic-rs/rustic#1032]
I would like to suggest `restic` implemented logic to try to keep the oldest snapshot

https://github.com/restic/restic/pull/4366

At the moment with `rustic` with for example forget policies:

```
[forget]
keep-last = 2
keep-daily = 2
keep-weekly = 2
keep-monthly = 2
```

$ rustic forget
using config ./rustic.toml
[INFO] repository rclone:OD01-ENC:restic-test: password is correct.
[INFO] using cache at /Users/kptsky/Library/Caches/rustic/xxxx
snapshots for (host [kp-laptop.local], label [], paths [/])

| ID       | Time                | Host            | Label | Tags        | Paths | Action | Reason  |
|----------|---------------------|-----------------|-------|-------------|-------|--------|---------|
| 005a026e | 2024-02-01 18:32:35 | kp-laptop.local |       | test_rustic | /     | keep   | last    |
|          |                     |                 |       |             |       |        | daily   |
|          |                     |                 |       |             |       |        | weekly  |
|          |                     |                 |       |             |       |        | monthly |
| f2c0df87 | 2024-01-31 15:09:11 | kp-laptop.local |       | test_rustic | /     | keep   | last    |
|          |                     |                 |       |             |       |        | daily   |
|          |                     |                 |       |             |       |        | monthly |
| 2690ada9 | 2024-01-31 04:41:52 | kp-laptop.local |       | test_rustic | /     | remove |         |
| 6fa4307a | 2024-01-30 08:05:13 | kp-laptop.local |       | test_rustic | /     | remove |         |
| 1dcc6ab7 | 2024-01-28 09:12:15 | kp-laptop.local |       | test_rustic | /     | keep   | weekly  |
| 2bc5f01c | 2023-12-31 20:44:27 | kp-laptop.local |       | test_rustic | /     | remove |         |

on the 1st of Feb 2024 the last snapshot from 31st of Dec is removed - it is only 31 days old....

I think it is something from `restic` that can be copied into `rustic` - much more logical IMO.

**Comment by aawsome** [comment]
@kapitainsky What do you expect as outcome here? I think restic does exactly the same here - there are 2 last, 2 daily, 2 weekly and 2 monthly snapshots kept...

**Comment by kapitainsky** [comment]
The latest `restic` would not remove the latest snapshot - see `restic` PR I linked

**Comment by kapitainsky** [comment]
And let's be pragmatic and logical - I want to keep last 2 monthly snapshot. In human terms it mean I do not expect 1 month snapshot to be removed (given it is the last one).

This is where pure algorithm logic does not work in real life.

**Comment by aawsome** [comment]
@kapitainsky Did you read my last comment?

 I do think that rustic behaves exactly like restic here and that restic also would remove that last snapshot. Simply because it keeps 2 monthly snapshots, as requested: The one from 2024-02-01 and the one from 2024-01-31.

Can you explain why you expect rustic to keep the snapshot from 2023-12-31 which would be the third monthly snapshot?

**Comment by kapitainsky** [comment]
Because between 2024-02-01 and  2024-01-31 there is 1 day difference? And you are dropping one from 2023-12-31? Which maybe satisfy some artificial logic of 2 months but defies any common sense.

Let's say I started my `rustic` backup on 31/DEC/2023 - Hurra! my data is safe now. And I set forget policies to keep last 2 months. So now I know that I can restore my files back up to 2 month. All was fine until 1/FEB/2024 when suddenly computer decided you now have only last two days of backups:) WOW. Because my 31/DEC/2023 backup (despite to be less than 2 months old) has been removed.

Very simple adjustment to prevent it is (like `restic`) to not to remove the oldest snapshot if it does satisfy set policies. It creates sort of exception to algorithmic removal process but makes it much more human friendly.

**Comment by aawsome** [comment]
If you want to keep monthly snapshots which are less then 2 month old, the `keep-within-monthly` option would be the right choice. Or, you tell rustic to keep 3 monthly snaphots knowing that in the beginning of a month the monthly snapshot is an early one.

> Very simple adjustment to prevent it is (like `restic`) to not to remove the oldest snapshot if it does satisfy set policies. It creates sort of exception to algorithmic removal process but makes it much more human friendly.

Please run `restic forget --dry-run`for your example. It will produce exactly the same result as rustic.

**Comment by aawsome** [comment]
I agree that this `keep-monthly` option may be irritating when you first start. But I decided to act exactly like restic here in order to not confuse people who run restic and rustic on the same repositories. The only thing I changed compared to restic was that rustic always keeps the oldest snapshots if there are not enough snapshots satisfying the `--keep-*` criteria. For restic, this was added with the mentioned PR about a half year ago. Now rustic and restic behave exactly identically.

**Comment by aawsome** [comment]
FWIW, the logic you are requesting and which has been added recently to restic has been added to rustic here:
https://github.com/rustic-rs/rustic/commit/bf21f8b1a351946be4331fd56428f1f9d6f9699c#diff-ffefaccfb096399d6312ab3946952e890c6cacb953210caed589cbf7609596aeR230
(snapshots are treated as candidates for a `keep-*` if they are either the oldest snapshot, the youngest snapshot or if the hour/day/week/.. doesn't match the last snapshot - Note that the oldest snapshot is in this list!)
